numba/cuda/cudaimpl.py

```diff
@@ -625,6 +625,40 @@ def ptx_atomic_sub(context, builder, dtype, ptr, val):
         return builder.atomic_rmw('sub', ptr, val, 'monotonic')
 
 
+@lower(stubs.atomic.inc, types.Array, types.intp, types.Any)
+@lower(stubs.atomic.inc, types.Array, types.UniTuple, types.Any)
+@lower(stubs.atomic.inc, types.Array, types.Tuple, types.Any)
+@_atomic_dispatcher
```

Comment:
Whilst the type restriction is documented, I think a load of these atomic functions could be typed more specifically to catch issues at type inference time. Example:

This is fine
```python
from numba import cuda
from numba import uint32
from numba import njit
import numpy as np

@cuda.jit('void(uint32[:], uint32, int8)')
def atomic_inc32(ary, idx, val):
    tid = cuda.threadIdx.x
    cuda.atomic.inc(ary, tid % 2, val)

ary = np.zeros(32, dtype=np.uint32)
idx = 0
rand_const = 100

atomic_inc32[1, 64](ary, idx, rand_const)
print(ary)
```
as `int8` is coercible to `uint32`. Whereas:
```python
from numba import cuda
from numba import uint32
from numba import njit
import numpy as np

@cuda.jit('void(uint32[:], uint32, complex64)')
def atomic_inc32(ary, idx, val):
    tid = cuda.threadIdx.x
    cuda.atomic.inc(ary, tid % 2, val)

ary = np.zeros(32, dtype=np.uint32)
idx = 0
rand_const = 100

atomic_inc32[1, 64](ary, idx, rand_const)
print(ary)
```
is not coercible and results in a lowering error. I don't think that this should be fixed in this PR, but perhaps this is something to look at in a follow up PR?

Reply:
Agree with catching issues at type inference time and a separate PR.